lib/OpenQA/WebSockets/Controller/Worker.pm

```diff
@@ -150,22 +150,23 @@ sub _message {
         log_trace "Received from worker $worker_id worker_status message: " . dumper($json)
           if LOG_WORKER_STATUS_MESSAGES;
 
+        my $workers = $schema->resultset('Workers');
+        my $worker = $workers->find($worker_id);
+
         # log that we 'saw' the worker
-        try {
-            $schema->txn_do(
-                sub {
-                    return undef unless my $w = $schema->resultset("Workers")->find($worker_id);
-                    log_debug("Updating seen of worker $worker_id from worker_status ($current_worker_status)");
-                    $w->seen({error => $current_worker_error});
-                });
+        if ($worker) {
+            try {
+                log_debug("Updating seen of worker $worker_id from worker_status ($current_worker_status)");
+                $worker->seen({error => $current_worker_error});
+            }
+            catch {
+                log_error("Failed updating seen and error status of worker $worker_id: $_");    # uncoverable statement
+            };
         }
-        catch {
-            log_error("Failed updating seen and error status of worker $worker_id: $_");    # uncoverable statement
-        };
 
         # send worker population
         try {
-            my $workers_population = $schema->resultset("Workers")->count();
+            my $workers_population = $workers->count();
```

Comment:
And if that whole worker population stuff gets to expensive I suggest to remove it completely (including the receiving side).

Reply:
That's exactly the plan for patch 3. 😁